gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,594 @@
+                            observation = env.reset()
+                pipe.send((observation, reward, done, info))
+            elif command == 'seed':
+                env.seed(data)
+                pipe.send(None)
+            elif command == 'close':
+                pipe.send(None)
+                break
+            elif command == '_call':
+                name, args, kwargs = data
+                if name in ['reset', 'step', 'seed', 'close']:
+                    raise ValueError('Trying to call function `{0}` with '
+                        '`_call`. Use `{0}` directly instead.'.format(name))
+                function = getattr(env, name)
+                if callable(function):
+                    pipe.send(function(*args, **kwargs))
+                else:
+                    pipe.send(function)
+            elif command == '_setattr':
+                name, value = data
+                setattr(env, name, value)
+                pipe.send(None)
+            elif command == '_check_observation_space':
+                pipe.send(data == env.observation_space)
+            elif command == '_restart':
+                observation = env.reset()
+                episode_done = True
+                if data == AsyncState.WAITING_RESET:
+                    pipe.send(observation)
+                elif data == AsyncState.WAITING_STEP:
+                    infos = {'AsyncVectorEnv.restart': True}
+                    pipe.send((observation, 0., True, infos))
+                elif data == AsyncState.WAITING_CALL:
+                    pipe.send(None)
+                else:
+                    raise NotImplementedError()
+            else:
+                raise RuntimeError('Received unknown command `{0}`. Must '
+                    'be one of {`reset`, `step`, `seed`, `close`, `_call`, '
+                    '`_setattr`, `_check_observation_space`, `_restart`}.'.format(command))
+    except Exception:
+        error_queue.put((index,) + sys.exc_info()[:2])
+        pipe.send(None)
+    finally:
+        env.close()
+
+
+def _worker_shared_memory(index, env_fn, pipe, parent_pipe, shared_memory,
+                          error_queue, episodic):
+    assert shared_memory is not None
+    env = env_fn()
+    observation_space = env.observation_space
+    _zero_observation = create_empty_array(observation_space, n=None, fn=np.zeros)
+    episode_done = False
+    parent_pipe.close()
+    try:
+        while True:
+            command, data = pipe.recv()
+            if command == 'reset':
+                observation = env.reset()
+                write_to_shared_memory(index, observation, shared_memory,
+                                       observation_space)
+                episode_done = False
+                pipe.send(None)
+            elif command == 'step':
+                if episodic and episode_done:
+                    observation = _zero_observation
+                    reward, done, info = 0., True, {}
```

Comment:
I think this can create subtle bugs downstream - for instance, a) if the environment produces something useful in info dictionary, and consumer relies on that, after episode end the consumer will complain about lack of right keys in info dictionary or b) reward / observation statistic computation can be skewed by this.

Reply:
I'd argue that the user shouldn't rely on specific keys in the info dictionary, nor the observation, after the episode ends because the results are undefined (as mentioned [here](https://github.com/openai/gym/blob/13f32c7689b9c8738d6db7f6b850ec216cb4d781/gym/core.py#L55)). But in the current state, I agree that reward and observation statistics might be skewed. This could be handled by having a key in the info dictionary for all steps after the episode ends (for now, `AsyncVectorEnv.end_episode` exists only for the final step of the episode)?